sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -499,8 +499,35 @@ case class TruncateTableCommand(
         val path = new Path(location.get)
         try {
           val fs = path.getFileSystem(hadoopConf)
+          val fileStatus = fs.getFileStatus(path)
+          // Not all fs impl. support these APIs.
+          val optPermission = Try(fileStatus.getPermission())
+          val optAcls = Try(fs.getAclStatus(path).getEntries)
+
           fs.delete(path, true)
```

Comment:
not familiar with Hadoop FS APIs, but do we have something like `rm -rf tablePath/*`. My point is, if we don't delete the parent folder, then we don't have this problem at all.

Reply:
In the [FileSystem API](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html), I didn't find one like that. 

But I just search again, there is [FileUtil API ](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileUtil.html#fullyDeleteContents-java.io.File-) that seems working like that.

I will test tomorrow to see if it work well to keep permission/acl in a Spark cluster like current approach.

Reply:
oh, I think `FileUtil.fullyDeleteContents` only works on local file (java.io.File), not for files on distributed file system like HDFS.

Reply:
So I think there is no single API to do `rm -rf tablePath/*` in Hadoop FS, if I haven't missed anything.

We can do `listStatus` and delete all contents in the given directory. But it is inefficient for DFS and is easier to have error.

Reply:
Do you know how Hive/Presto implement TRUNCATE TABLE? Are there other file attributes we need to retain?

Reply:
For Hive, I try to trace the code path for TRUNCATE TABLE:

1. HiveMetaStore. truncateTableInternal:

Get the locations of table/partitions and call Warehouse.deleteDir for each.

https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L3109

2. Warehouse.deleteDir delegates to MetaStoreFS.deleteDir.
3. HiveMetaStoreFsImpl (implements MetaStoreFS) delegates to FileUtils.moveToTrash which calls `FileSystem.delete(f, true)`. 

In `HiveMetaStore.truncateTableInternal`, after the location is deleted, new directory is created and previous file status including permissions, group, and ACLs are set to new directory.

Reply:
Now I retain permission and ACLs. I was doing `fs.setOwner` in first commit for retaining path owner and group. But `fs.setOwner` throws exception because it is only doable by super user.

In Hive code, since it is code running the metastore server, I think it is running with enough permission to set owner/group. For Spark, we may not running it with super user.

Reply:
In Presto, looks like it takes the approach by listing all files/directories under the path to delete (i.e., `rm -rf tablePath/*`), and recursively deleting them:

https://github.com/prestodb/presto/blob/9c21aaa659f9d4927fe12bec4b8015c8cbf4722e/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java#L485

https://github.com/prestodb/presto/blob/9c21aaa659f9d4927fe12bec4b8015c8cbf4722e/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java#L1796

Reply:
the presto way seems safer to me.

Reply:
I was concerned with the presto way on performance regression. For a table which has many files/directories, deleting them one by one could be a bottleneck.

If we add a config for retaining permission/ACLs when truncating table? Users can choose to disable it and directly delete top directory (current behavior) without retaining permission/ACLs.